Sources/Amplitude/Amplitude.m

```diff
@@ -780,21 +711,56 @@ - (void)uploadEventsWithLimit:(int)limit {
             return;
         }
 
-        long eventCount = [self.dbHelper getTotalEventCount];
+        long eventCount = [self->_eventsBuffer count];
         long numEvents = limit > 0 ? fminl(eventCount, limit) : eventCount;
         if (numEvents == 0) {
             self->_updatingCurrently = NO;
             [self endBackgroundTaskIfNeeded];
             return;
         }
-        NSMutableArray *events = [self.dbHelper getEvents:-1 limit:numEvents];
-        NSMutableArray *identifys = [self.dbHelper getIdentifys:-1 limit:numEvents];
-        NSDictionary *merged = [self mergeEventsAndIdentifys:events identifys:identifys numEvents:numEvents];
+        
+        long identifyCount = [self->_identifyBuffer count];
+        long numIdentify = limit > 0 ? fminl(identifyCount, limit) : identifyCount;
+        
+        NSMutableArray *events = [[self->_eventsBuffer subarrayWithRange:NSMakeRange(0, numEvents)] mutableCopy];
+        NSMutableArray *identifys = [[self->_identifyBuffer subarrayWithRange:NSMakeRange(0, numIdentify)] mutableCopy];
+        NSDictionary *merged = [self mergeEventsAndIdentifys:events identifys:identifys numEvents:(numEvents+numIdentify)];
 
+        for (NSDictionary *event in events) {
+            // convert event dictionary to JSON String
+            NSError *error = nil;
+            NSData *jsonData = [NSJSONSerialization dataWithJSONObject:[AMPUtils makeJSONSerializable:event] options:0 error:&error];
+            if (error != nil) {
+                AMPLITUDE_ERROR(@"ERROR: could not JSONSerialize event type: %@", error);
+                continue;
+            }
+            NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+            if ([AMPUtils isEmptyString:jsonString]) {
+                AMPLITUDE_ERROR(@"ERROR: JSONSerializing event type event resulted in an NULL string");
+                continue;
+            }
+            [AMPStorage storeEvent:jsonString];
```

Comment:
So we are changing from storeEvent on event creation, and we save the events to file before upload happens. Is there any difference or performance improvements we can see by this way? And with this change, we just rely on memory more and can still lose events before we call upload?

Reply:
Yes, there's better improvements for doing it as a batch. It's better to write to disk at once than many times elsewhere. Maybe this approach could go back to streaming if we used NSOutputStream. I got really good results using a batch, see the slack chat where I zipped some time profiles.